src/main/java/org/dbos/apiary/xa/MySQLXAConnection.java

```diff
@@ -0,0 +1,198 @@
+        } catch(Exception exception) {
+            logger.info(exception.getMessage());
+            throw exception;
+        }
+    }
+
+    public void XAEnd(ApiaryXID xid) throws Exception {
+        try {
+            this.xaconnection.get().getXAResource().end(xid, javax.transaction.xa.XAResource.TMSUCCESS);
+        } catch(Exception exception) {
+            logger.info(exception.getMessage());
+            throw exception;
+        }
+    }
+
+    public void XARollback(ApiaryXID xid) throws Exception {
+        try {
+            this.xaconnection.get().getXAResource().rollback(xid);
+        } catch(Exception exception) {
+            logger.info(exception.getMessage());
+            throw exception;
+        }
+    }
+
+    public boolean XAPrepare(ApiaryXID xid) throws Exception {
+        try {
+            int res = this.xaconnection.get().getXAResource().prepare(xid);
+            if (res == javax.transaction.xa.XAResource.XA_OK || res == javax.transaction.xa.XAResource.XA_RDONLY) {
+                return true;
+            }
+        } catch(Exception exception) {
+            logger.info(exception.getMessage());
+            throw exception;
+        }
+        return false;
+    }
+
+    public void XACommit(ApiaryXID xid) throws Exception  {
+        try {
+            this.xaconnection.get().getXAResource().commit(xid, false);
+        } catch(Exception exception) {
+            logger.info(exception.getMessage());
+        }
+    }
+
+    private void prepareStatement(PreparedStatement ps, Object[] input) throws SQLException {
+        for (int i = 0; i < input.length; i++) {
+            Object o = input[i];
+            if (o instanceof Integer) {
+                ps.setInt(i + 1, (Integer) o);
+            } else if (o instanceof String) {
+                ps.setString(i + 1, (String) o);
+            } else {
+                assert (false); // TODO: More types.
+            }
+        }
+    }
+
+    /**
+     * Execute a database update.
+     * @param procedure a SQL DML statement (e.g., INSERT, UPDATE, DELETE).
+     * @param input     input parameters for the SQL statement.
+     */
+    public void executeUpdate(String procedure, Object... input) throws SQLException {
+        // First, prepare statement. Then, execute.
+        PreparedStatement pstmt = this.connection.get().prepareStatement(procedure);
+        prepareStatement(pstmt, input);
+        pstmt.executeUpdate();
+    }
+
+    /**
+     * Execute a database query.
+     * @param procedure a SQL query.
+     * @param input     input parameters for the SQL statement.
+     */
+    public ResultSet executeQuery(String procedure, Object... input) throws SQLException {
+        PreparedStatement pstmt = this.connection.get().prepareStatement(procedure, ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
+        prepareStatement(pstmt, input);
+        ResultSet rs = pstmt.executeQuery();
+        return rs;
+    }
```

Comment:
There seems to be a lot of repetition between MySQL and Postgres code. Maybe you could put some of those in the base `XADBConnection` class. Because most of the JDBC interface between those two DBMSs is the same.